File: posx/api/xz_report.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
import frappe
from toolz.curried import merge


@frappe.whitelist()
def create_opening(
    opening_cash, company, pos_profile, user=None, branch=None, start_datetime=None
):
    xzreport = frappe.get_doc(
        {
            "doctype": "XZ Report",
            "start_datetime": start_datetime or frappe.utils.now(),
            "user": user or frappe.session.user,
            "pos_profile": pos_profile,
            "branch": branch,
            "company": company,
            "opening_cash": opening_cash,
        }
    ).insert(ignore_permissions=True, ignore_mandatory=True)
    return xzreport.name


@frappe.whitelist()
def get_unclosed(user, pos_profile=None, company=None):
    return frappe.db.exists(
        "XZ Report",
        merge(
            {"user": user, "company": company, "docstatus": 0},
            {"pos_profile": pos_profile} if pos_profile else {},
        ),
    )

File: posx/public/js/pos/xz_report.js
import makeExtension from '../utils/make-extension';

export async function get_xz_report(pos_profile, company) {
  const { message: { px_enable_xz_report: enable_xz_report } = {} } =
    (await frappe.db.get_value(
      'POS Profile',
      pos_profile,
      'px_enable_xz_report'
    )) || {};

  if (!enable_xz_report) {
    return null;
  }

  const { message: xz_report } = await frappe.call({
    method: 'posx.api.xz_report.get_unclosed',
    args: { user: frappe.session.user, pos_profile, company },
  });
  if (xz_report) {
    return xz_report;
  }

  const dialog = new frappe.ui.Dialog({
    title: __('Enter Opening Cash'),
    fields: [
      {
        fieldtype: 'Datetime',
        fieldname: 'start_datetime',
        label: __('Start Datetime'),
        default: frappe.datetime.now_datetime(),
      },
      { fieldtype: 'Column Break' },
      {
        fieldtype: 'Currency',
        fieldname: 'opening_cash',
        label: __('Amount'),
      },
    ],
  });
  dialog.get_close_btn().hide();
  dialog.show();

  return new Promise((resolve, reject) => {
    dialog.set_primary_action('Enter', async function () {
      try {
        const { start_datetime, opening_cash } = dialog.get_values();
        const { message: xz_report } = await frappe.call({
          method: 'posx.api.xz_report.create_opening',
          args: { start_datetime, opening_cash, company, pos_profile },
        });
        if (!xz_report) {
          throw new Error(__('Unable to create XZ Report opening entry.'));
        }
        resolve(xz_report);
      } catch (e) {
        frappe.msgprint({
          message: e.message,
          title: __('Warning'),
          indicator: 'orange',
        });
        reject(e);
      } finally {
        dialog.hide();
        dialog.$wrapper.remove();
      }
    });
  });
}

export default function xz_report(Pos) {
  return makeExtension(
    'xz_report',
    class PosWithXzReport extends Pos {
      async set_pos_profile_data() {
        const result = await super.set_pos_profile_data();
        if (this.frm.config.px_enable_xz_report) {
          this._update_menu();
        }
        return result;
      }
      _update_menu() {
        this.page.menu
          .find(`a.grey-link:contains("${__('Close the POS')}")`)
          .parent()
          .hide();
        this.page.add_menu_item(
          'XZ Report',
          async function () {
            const { pos_profile, company } = this.frm.doc;
            const xz_report = await get_xz_report(pos_profile, company);
            frappe.set_route('Form', 'XZ Report', xz_report);
          }.bind(this)
        );
      }
    }
  );
}
